etna/ciim/utils.py

```diff
@@ -14,6 +14,58 @@ def underscore_to_camelcase(word, lower_first=True):
     return result
 
 
+class ValueExtractionError(Exception):
+    pass
+
+
+def extract(
+    source: Dict[str, Any], key: str, default: Any = "", fail_loudly: bool = False
+) -> Any:
+    """
+    Attempts to extract `key` (a string with multiple '.' to indicate
+    traversal) from `source` (a complex multi-level dict where values may
+    by lists or other complex types) and return the value.
+
+    Raises ``ValueExtractionError` when the extraction is unsuccessful.
+
+    Implementation inspired by ``django.template.base.Variable._resolve_lookup```:
+    https://github.com/django/django/blob/0dd29209091280ccf34e07c9468746c396b7778e/django/template/base.py#L867-L942
+    """
+    current = source
+    lookups = tuple(key.split("."))
+
+    try:
+        for bit in lookups:
+            try:  # dictionary lookup
+                current = current[bit]
+                # ValueError/IndexError are for numpy.array lookup on
+                # numpy < 1.9 and 1.9+ respectively
+            except (TypeError, AttributeError, KeyError, ValueError, IndexError):
+                try:  # attribute lookup
+                    current = getattr(current, bit)
```

Comment:
`current = getattr(current, bit)` Could not find a scenario satisfying this assignment, Can we a test case for it?

Reply:
@TNA-Allan This happens whenever the initial dict lookup fails. In the tests, you'll see that we're looking up attributes from date objects in lots of places... that is where this code will be working